doc/secure_boot/hld_secure_boot.md

```diff
@@ -0,0 +1,383 @@
+
+SONiC is using the initrd, so we will remain it as in the original repo unsigned. (In order to sign it is required to use a gpg key, but there is not really a use case for that. (Probably future work.)
+
+Grub.cfg - The grub.cfg file is the GRUB configuration file. It is generated by the grub2-mkconfig program using a set of primary configuration files and the grub default file as a source for user configuration specifications. The /boot/grub2/grub.cfg file is first generated during Linux installation and regenerated when a new kernel is installed.
+
+The grub.cfg file contains Bash-like code and a list of installed kernels in an array ordered by sequence of installation. For example, if you have four installed kernels, the most recent kernel will be at index 0, the previous kernel will be at index 1, and the oldest kernel will be at index 3.
+
+(Not modification required)
+
+##### First instance of trust
+SB chain of trust has different options for this first instance of trust.
+This HLD show 2 options:
+1.  BIOS/UEFI as the first instance of trust, vendors can store their public keys in the UEFI DB, by doing that, they enforce to boot images that were only signed with a private key that matches the vendor's public key.
+the diagram can be founded in the "Run-time verification flow diagram with BIOS as the first instance of trust".
+This is a generic way that Linux support using the secure boot feature and we add support to it as well.
+Linux Debian SB link: https://wiki.debian.org/SecureBoot
+
+2. HW as a root of trust, full HW root of trust is reliant on the HW vendor and production flow since customers can add this by themselves depending on the HW platform they have.
+So in this HLD, we showed the diagram "Run-time verification flow diagram with HW root of trust", but, this implementation should be decided by the vendor.
+
+An example of HW root of trust can be the intel boot guard, pls see the info in the links below:
+
+- https://www.intel.com/content/dam/www/central-libraries/us/en/documents/below-the-os-security-white-paper.pdf
+
+- https://edk2-docs.gitbook.io/understanding-the-uefi-secure-boot-chain/secure_boot_chain_in_uefi/intel_boot_guard
+
+
+####  1.7.3. <a name='SignFlowdiagramdescription:'></a>Sign Flow diagram description:
+Sign flow occurs when building the SONiC image.
+In the Build process, we need to sign the components shown in the Design Diagram chapter (shim, grub, vmlinuz, kernel modules) and do some modifications in the Linux kernel, that in order to pass the runtime secure boot verification.
+
+This chapter will be described which SONiC modification is required.
+
+Compilation Flag added:
+- SECURE_BOOT_BUILD – have 3 options
+  - no_sign  (default)
+  - dev
+  - release
+- SECURE_BOOT_DEV_KEY_PATH – have the path where dev private efi key stored
```

Comment:
Can we use the existing variables? SIGNING_KEY, SIGNING_CERT, see https://github.com/sonic-net/sonic-buildimage/pull/10557/files

Is it possible to use the same variables for production and dev, but different keys?

Reply:
I'll answer - using the same variables for different features can cause us bugs in the future (we cannot be certain that these certificates and keys are being used in the same manner and in the same format).
If you have the notion that these items will be identical by 100% - we can consider this. I, for one, am not 100% sure on this matter.

Reply:
We can have a different feature flag. Do we have to use the different the certificate config?
It will be great if we can use the existing config variables which are also for secure boot, it reduces the config complexity, make the user easy to do the certificate config when using secure boot feature.

Reply:
what do you mean exactly by "certificate config".

Like mentioned in the comment before, we are using a different build flag because our secure boot flow is different, we are signing shim, grub, vmlinuz and Kernel Modules with specific algos and keys (described in our HLD).
The existing flag is singing just vmlinuz and they can use any algo and key, so in order to avoid conflicts it's better to save the configurations and key parameters of the feature separately, in addition, do not use the features simultaneously. (I added a condition that supporting  to use our feature just if the other feature is off. (see in build_debian.sh) )